File: js/ControlDigit/.solved/Upc.js
import { getDigitsFromLeastSignificant, sumWithWeights } from './ControlDigit'

const weights = createWeights();

export default function upc (number) {
    const digits = getDigitsFromLeastSignificant(number);
    const m = sumWithWeights(digits, weights) % 10;
    return m === 0 ? 0 : 10 - m;
}

function createWeights () {
    const generator = repeat([3, 1]);
    return Array.from(new Array(11), () => generator.next().value);
}

function* repeat (items) {
    while (true) {
        for (let i = 0; i < items.length; i++) {
            yield items[i];
        }
    }
}

File: python/ControlDigit/Solved/Upc/upc.py
from control_digit import get_digits_from_least_significant, repeat, sum_with_weights


def upc(number: int) -> int:
    weights = repeat([3, 1])
    digits = get_digits_from_least_significant(number)
    m = sum_with_weights(digits, weights) % 10
    return 0 if m == 0 else 10 - m
